### PR TITLE
UI-500 UI-501 UI-516: Form, FormElement, Toggle

### PR DESCRIPTION
**Pull request by hanzge**
- Refactor + unit test for Form, FormElement and Toggle
- Rewrite storybooks for Form
- Fix Input bug by adding `inputRef` to its `ref`
- Remove `label` props from Checkbox, Radio.

Breaking change of the form pattern:
Before:
```
<Form>
  <Input label="Email" />
</Form>
```

Now:
```
<Form>
  <FormElement>
    <Label>Email</Label>
    <Input />
  </FormElement>
</Form>
```

### CODE REVIEW

File: lib/components/Form/FormElement.js [hunk 1]
```diff
@@ -2,10 +2,41 @@ import React from 'react';
...
+  cols: PropTypes.number,
+  error: PropTypes.oneOfType([
+    PropTypes.bool,
+    PropTypes.string,
+    PropTypes.shape({
+      message: PropTypes.string,
+    }),
+  ]),
+  formElementRef: PropTypes.func,
+  id: PropTypes.string,
```
Comment (joshsylvester) [inline lib/components/Form/FormElement.js]: based on previous conversations - do we want this here?

File: lib/components/Form/FormElement.js [hunk 2]
```diff
@@ -2,10 +2,41 @@ import React from 'react';
...
+  isRequired: PropTypes.bool,
+  label: PropTypes.string,
+  readOnly: PropTypes.bool,
```
Comment (joshsylvester) [inline lib/components/Form/FormElement.js]: `isReadOnly`

File: lib/components/Form/FormElement.js [hunk 3]
```diff
@@ -14,107 +45,80 @@ export default class FormElement extends React.Component {
...
-      dropdown, className, totalCols, cols, error,
-      children, style, ...props
-    } = this.props;
+    const { className, totalCols, cols, error, children, ...props } = this.props;
     const labelElem = this.renderLabel();
-    const controlElem = this.renderControl({ children, dropdown, error });
+    const controlElem = this.renderControl();
     const formElemChildren = [labelElem, controlElem];
     return this.renderFormElement({
       ...props,
```
Comment (joshsylvester) [inline lib/components/Form/FormElement.js]: shouldn't the `...props` be at the end?

File: stories/ToggleStories.js [hunk 4]
```diff
@@ -3,26 +3,64 @@ import { storiesOf } from '@storybook/react';
 import { withInfo } from '@storybook/addon-info';
 import { action } from '@storybook/addon-actions';
 import { text, boolean } from '@storybook/addon-knobs';
-import { Toggle } from './../lib';
+import { Toggle } from '../lib';
+
+class ToggleWrapper extends React.Component {
```
Comment (joshsylvester) [inline stories/ToggleStories.js]: If we're adding this here - why isn't this in the toggle component itself? shouldn't the toggle know it's own state?

File: lib/components/Form/FormElement.js [hunk 5]
```diff
@@ -2,10 +2,41 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import classnames from 'classnames';
 
-export default class FormElement extends React.Component {
-  renderFormElement(props) {
-    const { className, error, totalCols, cols = 1, formElementRef, children } = props;
+const propTypes = {
+  children: PropTypes.node,
```
Comment (mallik-svmx) [inline lib/components/Form/FormElement.js]: Let's continue to add the prop comments, useful part of documentation. (I know a few props are common across components, but remaining once are unique that's helpful)

File: lib/components/Form/FormElement.js [hunk 6]
```diff
@@ -14,107 +45,80 @@ export default class FormElement extends React.Component {
...
-          }
-        </label> :
-        undefined
-    );
-  }
+  renderLabel = () => {
+    const { id, label, isRequired } = this.props;
+    return label ? (
+      <label
+        key="form-element-label"
```
Comment (mallik-svmx) [inline lib/components/Form/FormElement.js]: key with just static string in this case may not be useful.

**Comment by hanzge** [comment]
This PR will be continued to work on after https://github.com/ServiceMax-Engineering/ui-components-lightning/pull/24 is merged.

File: lib/components/Form/FormElement.js [hunk 7]
```diff
@@ -2,10 +2,41 @@ import React from 'react';
...
+
+const defaultProps = {
+  cols: 1,
+  error: false,
+  formElementRef: null,
+  id: '',
+  isRequired: false,
+  label: '',
+  readOnly: false,
+  totalCols: 1,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 8]
```diff
@@ -14,107 +45,80 @@ export default class FormElement extends React.Component {
...
-    const controlElem = this.renderControl({ children, dropdown, error });
+    const controlElem = this.renderControl();
     const formElemChildren = [labelElem, controlElem];
     return this.renderFormElement({
       ...props,
+      children: formElemChildren,
       className,
+      cols,
       error,
       totalCols,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 9]
```diff
@@ -2,10 +2,41 @@ import React from 'react';
...
+    PropTypes.shape({
+      message: PropTypes.string,
+    }),
+  ]),
+  formElementRef: PropTypes.func,
+  id: PropTypes.string,
+  isRequired: PropTypes.bool,
+  label: PropTypes.string,
+  readOnly: PropTypes.bool,
+  totalCols: PropTypes.number,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 10]
```diff
@@ -14,107 +45,80 @@ export default class FormElement extends React.Component {
...
+      </span>
+    );
+  };
 
   render() {
-    const {
-      dropdown, className, totalCols, cols, error,
-      children, style, ...props
-    } = this.props;
+    const { className, totalCols, cols, error, children, ...props } = this.props;
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: MAJOR Remove this useless assignment to local variable "children" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

**Comment by maxio** [comment]
Test FAILed.
Refer to this link for build results: https://jenkins.servicemax.io/job/UIComponents/job/ui-components-lightning/3/.

File: lib/components/Form/Form.js [hunk 11]
```diff
@@ -1,48 +1,35 @@
...
+const propTypes = {
+  children: PropTypes.node,
+  /**
+   * Custom classNames besides default ones
+   */
   className: PropTypes.string,
+  /**
+   * Form layout choice
+   */
   type: PropTypes.oneOf(FORM_TYPES),
```
Comment (maxsonar) [inline lib/components/Form/Form.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/Form.js [hunk 12]
```diff
@@ -1,48 +1,35 @@
...
-  children: PropTypes.node,
 };
 
-Form.defaultProps = {
+const defaultProps = {
   type: 'stacked',
```
Comment (maxsonar) [inline lib/components/Form/Form.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/Form.js [hunk 13]
```diff
@@ -1,48 +1,35 @@
...
 };
+
+const Form = ({ className, type, children, ...props }) => {
```
Comment (maxsonar) [inline lib/components/Form/Form.js]: MINOR Rename this 'Form' function to match the regular expression ^[a-z][a-zA-Z0-9]*$ [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS100)

File: lib/components/Checkbox/CheckboxGroup.js [hunk 14]
```diff
@@ -38,7 +34,6 @@ const defaultProps = {
   children: [],
   className: '',
   cols: 1,
-  label: '',
   name: '',
   onChange: () => {},
   totalCols: 1,
```
Comment (maxsonar) [inline lib/components/Checkbox/CheckboxGroup.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 15]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
+   */
+  formElementRef: PropTypes.func,
+  /**
+   * Read-only form element
+   */
+  isReadOnly: PropTypes.bool,
+  /**
+   * Total number of columns
+   */
+  totalCols: PropTypes.number,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 16]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
+   */
+  cols: PropTypes.number,
+  /**
+   * Error message
+   */
+  error: PropTypes.oneOfType([
+    PropTypes.bool,
+    PropTypes.string,
+    PropTypes.shape({
+      message: PropTypes.string,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 17]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
+    }),
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 18]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
-    const { readOnly } = this.props;
+class FormElement extends React.Component {
+  renderControl = () => {
+    const { children, error, isReadOnly } = this.props;
     const formElementControlClassNames = classnames(
+      'FormElement__control',
       'slds-form-element__control',
-      { 'slds-has-divider--bottom': readOnly },
+      {
+        'slds-has-divider--bottom': isReadOnly,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 19]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
-    const formElemChildren = [labelElem, controlElem];
-    return this.renderFormElement({
-      ...props,
+    const { className, totalCols, cols, error, children, formElementRef } = this.props;
+    const formElementClassNames = classnames(
+      'FormElement',
+      'slds-form-element',
+      {
+        'slds-has-error': error,
+        [`slds-size--${cols}-of-${totalCols}`]: typeof totalCols === 'number',
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Form/FormElement.js [hunk 20]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
-        undefined
-    );
-  }
+const defaultProps = {
+  cols: 1,
+  error: false,
+  formElementRef: null,
+  isReadOnly: false,
+  isRequired: false,
+  totalCols: null,
```
Comment (maxsonar) [inline lib/components/Form/FormElement.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Toggle/Toggle.js [hunk 21]
```diff
@@ -1,58 +1,80 @@
...
+    );
+  };
+
+  render() {
+    const {
+      id = this.defaultId,
+      className,
+      disabledText,
+      enabledText,
+      defaultChecked,
```
Comment (maxsonar) [inline lib/components/Toggle/Toggle.js]: MAJOR Remove this useless assignment to local variable "defaultChecked" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

File: lib/components/Toggle/Toggle.js [hunk 22]
```diff
@@ -1,58 +1,80 @@
...
+      isDisabled,
+      onChange,
```
Comment (maxsonar) [inline lib/components/Toggle/Toggle.js]: MAJOR Remove this useless assignment to local variable "onChange" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

File: lib/components/Toggle/Toggle.js [hunk 23]
```diff
@@ -1,58 +1,80 @@
...
+};
+
+const defaultProps = {
+  defaultChecked: false,
+  disabledText: 'Disabled',
+  enabledText: 'Enabled',
+  id: undefined,
+  isDisabled: false,
+  onChange: null,
+  value: '',
```
Comment (maxsonar) [inline lib/components/Toggle/Toggle.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Toggle/Toggle.js [hunk 24]
```diff
@@ -1,58 +1,80 @@
...
+};
+
+class Toggle extends Component {
+  state = { isChecked: this.props.defaultChecked };
```
Comment (maxsonar) [inline lib/components/Toggle/Toggle.js]: CRITICAL Remove the use of "this". [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS2990)

File: lib/components/Toggle/Toggle.js [hunk 25]
```diff
@@ -1,58 +1,80 @@
...
-    const toggleClassNames = classnames(className, 'slds-checkbox--toggle slds-grid');
+const propTypes = {
+  className: PropTypes.string,
+  defaultChecked: PropTypes.bool,
+  disabledText: PropTypes.string,
+  enabledText: PropTypes.string,
+  id: PropTypes.string,
+  isDisabled: PropTypes.bool,
+  onChange: PropTypes.func,
+  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
Comment (maxsonar) [inline lib/components/Toggle/Toggle.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

File: lib/components/Input/InputField.js [hunk 26]
```diff
@@ -54,14 +54,22 @@ class InputField extends Component {
...
-  }
+  };
 
   render() {
     const {
-      className, bare, value, onKeyDown, onValueChange,
+      className,
+      bare,
+      value,
+      onKeyDown,
```
Comment (maxsonar) [inline lib/components/Input/InputField.js]: MAJOR Remove this useless assignment to local variable "onKeyDown" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

File: lib/components/Input/InputField.js [hunk 27]
```diff
@@ -54,14 +54,22 @@ class InputField extends Component {
...
+      onValueChange,
```
Comment (maxsonar) [inline lib/components/Input/InputField.js]: MAJOR Remove this useless assignment to local variable "onValueChange" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

File: lib/components/Input/InputField.js [hunk 28]
```diff
@@ -54,14 +54,22 @@ class InputField extends Component {
...
+      inputRef,
+      symbolPattern,
```
Comment (maxsonar) [inline lib/components/Input/InputField.js]: MAJOR Remove this useless assignment to local variable "symbolPattern" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

File: lib/components/Input/InputField.js [hunk 29]
```diff
@@ -54,14 +54,22 @@ class InputField extends Component {
...
     }
-  }
+  };
 
   render() {
     const {
-      className, bare, value, onKeyDown, onValueChange,
+      className,
+      bare,
+      value,
```
Comment (maxsonar) [inline lib/components/Input/InputField.js]: MAJOR Remove this useless assignment to local variable "value" [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3AS1854)

File: lib/components/Radio/RadioGroup.js [hunk 30]
```diff
@@ -34,7 +30,6 @@ const defaultProps = {
   children: null,
   className: '',
   cols: 1,
-  label: '',
   totalCols: 1,
```
Comment (maxsonar) [inline lib/components/Radio/RadioGroup.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

**Comment by maxsonar** [comment]
SonarQube analysis reported 31 issues
* ![BLOCKER][BLOCKER] 22 blocker
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [CheckboxGroup.js#L30](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Checkbox/CheckboxGroup.js#L30): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [CheckboxGroup.js#L74](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Checkbox/CheckboxGroup.js#L74): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L19](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Input/InputField.js#L19): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L26](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Input/InputField.js#L26): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L32](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Input/InputField.js#L32): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L40](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Input/InputField.js#L40): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [Label.js#L18](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Label/Label.js#L18): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [Label.js#L23](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Label/Label.js#L23): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [RadioGroup.js#L26](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Radio/RadioGroup.js#L26): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [RadioGroup.js#L40](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/df01a2da8486cb0e324bfba27466ec66cc33779d/lib/components/Radio/RadioGroup.js#L40): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by maxio** [comment]
Test FAILed.
Refer to this link for build results: https://jenkins.servicemax.io/job/UIComponents/job/ui-components-lightning/7/.

File: lib/components/Toggle/Toggle.js [hunk 31]
```diff
@@ -1,58 +1,80 @@
...
 import classnames from 'classnames';
-import { FormElement } from './../Form';
+import { uuid } from '../utils';
 
-export default class Toggle extends Component {
-  renderToggle({ className, label, ...props }) {
-    const toggleClassNames = classnames(className, 'slds-checkbox--toggle slds-grid');
+const propTypes = {
+  className: PropTypes.string,
+  defaultChecked: PropTypes.bool,
```
Comment (sheela-svmx) [inline lib/components/Toggle/Toggle.js]: Adding comments here? for Proptypes?

File: lib/components/Form/FormElement.js [hunk 32]
```diff
@@ -2,119 +2,105 @@ import React from 'react';
...
-        ref={formElementRef}
-        key="form-element"
-        className={formElementClassNames}
-      >
-        { children }
-      </div>
-    );
-  }
+const propTypes = {
+  children: PropTypes.node,
```
Comment (sheela-svmx) [inline lib/components/Form/FormElement.js]: I know its a pain, but lets keep it consistent by adding comment to all proptypes

File: lib/components/Toggle/Toggle.js (hunk 31 above)
Comment (sheela-svmx) [reply lib/components/Toggle/Toggle.js]: If this is outside the scope of your work , dont bother, since i saw you modify this file , i left that comment.

Comment (hanzge) [reply lib/components/Toggle/Toggle.js]: Sure, I'm adding it

File: lib/components/Toggle/Toggle.js [hunk 33]
```diff
@@ -1,58 +1,104 @@
...
+   */
+  isDisabled: PropTypes.bool,
+  /**
+   * Callback when changing Toggle status
+   */
+  onChange: PropTypes.func,
+  /**
+   * Value of the Toggle
+   */
+  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
Comment (maxsonar) [inline lib/components/Toggle/Toggle.js]: BLOCKER Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

**Comment by maxsonar** [comment]
SonarQube analysis reported 31 issues
* ![BLOCKER][BLOCKER] 22 blocker
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [CheckboxGroup.js#L30](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Checkbox/CheckboxGroup.js#L30): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [CheckboxGroup.js#L74](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Checkbox/CheckboxGroup.js#L74): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L19](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Input/InputField.js#L19): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L26](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Input/InputField.js#L26): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L32](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Input/InputField.js#L32): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [InputField.js#L40](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Input/InputField.js#L40): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [Label.js#L18](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Label/Label.js#L18): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [Label.js#L23](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Label/Label.js#L23): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [RadioGroup.js#L26](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Radio/RadioGroup.js#L26): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)
1. ![BLOCKER][BLOCKER] [RadioGroup.js#L40](https://github.com/ServiceMax-Engineering/ui-components-lightning/blob/b8929169526a6061ec0d21459b43e6cb315f995c/lib/components/Radio/RadioGroup.js#L40): Remove this trailing comma. [rule](https://sonarqube.servicemax.io/coding_rules#rule_key=javascript%3ATrailingComma)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by maxio** [comment]
Test FAILed.
Refer to this link for build results: https://jenkins.servicemax.io/job/UIComponents/job/ui-components-lightning/9/.

File: lib/components/Input/InputField.js (hunk 29 above)
Comment (joshsylvester) [reply lib/components/Input/InputField.js]: this comment is accurate - there is no point to pull `value` (and a few other props here) out of `this.props` if you're not going to use them.

File: lib/components/Toggle/Toggle.js (hunk 24 above)
Comment (joshsylvester) [reply lib/components/Toggle/Toggle.js]: shouldn't the `state` be initialized in the `constructor(props) {}` method?

File: stories/FormStories.js [hunk 34]
```diff
@@ -2,129 +2,192 @@ import React from 'react';
...
+  Textarea,
+  Checkbox,
+  CheckboxGroup,
+  Label,
+  Radio,
+  RadioGroup,
+  Select,
+  Option,
+  Button,
+  FormElement,
```
Comment (joshsylvester) [inline stories/FormStories.js]: alphabetical?

Comment (hanzge) [reply stories/FormStories.js]:
```js
render() {
    const {
      className,
      bare,
      value,
      onKeyDown,
      onValueChange,
      inputRef,
      symbolPattern,
      ...inputProps
    } = this.props;
    const inputClassNames = classnames(
      className,
      'Input__field',
      bare ? 'slds-input--bare' : 'slds-input',
    );
    return (
      <input
        className={inputClassNames}
        onChange={event => {
          this.props.onValueChange.call(this, event);
          this.onChange(event);
        }}
        onKeyDown={this.onKeyDown}
        ref={inputRef}
        value={this.state.value}
        {...inputProps} //inputProps, symbolPattern shouldn't be here.
      />
    );
  }
...

Comment (hanzge) [reply stories/FormStories.js]: Never mind. I think I got your point about the InputField props.

**Pull request merged**